Lib/typing.py

```diff
@@ -228,25 +228,41 @@ def _remove_dups_flatten(parameters):
     return tuple(params)
 
 
+def _flatten_literal_params(parameters):
+    """An internal helper for Literal creation: flatten Literals among parameters"""
+    params = []
+    for p in parameters:
+        if isinstance(p, _LiteralGenericAlias):
+            params.extend(p.__args__)
+        else:
+            params.append(p)
+    return tuple(params)
+
+
 _cleanups = []
 
 
-def _tp_cache(func):
+def _tp_cache(typed=False):
     """Internal wrapper caching __getitem__ of generic types with a fallback to
     original function for non-hashable arguments.
```

Comment:
Add a note that this supports both `@tp_cache` and `@tp_cache(typed=True)`.

(And honestly I think there are nicer ways to write the logic for that. Maybe this works?
```
def _tp_cache(func=None, /, *, typed=False):
    if func is not None:
        return _tp_cache(typed=False)(func)
```
)

Reply:
You are right, such `_tp_cache` signature looks much cleaner.

There are so many ways to make a decorator with parameters. It will be great to have some kind of decorator at `functools` which will allow us to easily write  decorators with parameters, it might look like this:
```py
from functools import decorator

@decorator
def _tp_cache(func=None, /, *, typed=False):
    """Internal wrapper caching __getitem__ of generic types with a fallback to
    original function for non-hashable arguments.
    """
    cached = functools.lru_cache(typed=typed)(func)
    _cleanups.append(cached.cache_clear)

    @functools.wraps(func)
    def inner(*args, **kwds):
        try:
            return cached(*args, **kwds)
        except TypeError:
            pass  # All real errors (not unhashable args) are raised below.
        return func(*args, **kwds)
    return inner
```

Should I create an issue regarding that?